### PR TITLE
Iterative statistics

### PR DESCRIPTION
**Pull request by mbaudin47**
This PR implements several iterative statistics:
- mean
- variance
- kurtosis
- skewness
- extrema
- threshold exceedance.
  
This PR follows the work started at PR #1902.

Three classes are created:
- IterativeExtrema
- IterativeMoments
- IterativeThresholdExceedance
 
Here is an example of `IterativeExtrema`:
```python
>>> import openturns as ot
>>> dim = 5
>>> myExtrema = ot.IterativeExtrema(dim)
>>> n = ot.Normal(dim)
>>> size = 50
>>> for i in range(size):
>>>     point = n.getRealization()
>>>     myExtrema.increment(point)
>>> print(myExtrema.getIterationNumber())
50
>>> print(myExtrema.getMin())
[-2.4067,-2.53986,-2.29006,-3.09737,-2.18139]
>>> print(myExtrema.getMax())
[3.01263,3.02799,1.85579,2.11968,1.36783]
```

Here is an example of `IterativeMoments`:
```python
>>> import openturns as ot
>>> distNormal = ot.Normal()
>>> order = 4
>>> dim = 1
>>> iterMoments = ot.IterativeMoments(order, dim)
>>> size = 2000
>>> meanEvolution = ot.Sample()
>>> for i in range(size):
>>>     point = distNormal.getRealization()
>>>     iterMoments.increment(point)
>>> print('Mean:     ', iterMoments.getMean())
Mean:      [-0.00726852]
>>> print('Variance: ', iterMoments.getVariance())
Variance:  [0.99156]
>>> print('Skewness: ', iterMoments.getSkewness())
Skewness:  [0.0577277]
>>> print('Kurtosis: ', iterMoments.getKurtosis())
Kurtosis:  [3.11128]
```

Here is an example of `IterativeThresholdExceedance`:
```python
>>> import openturns as ot
>>> dim = 5
>>> threshold = 0.5
>>> myThresholdExceedance = ot.IterativeThresholdExceedance(dim, threshold)
>>> n = ot.Normal(dim)
>>> size = 50
>>> for i in range(size):
>>>     point = n.getRealization()
>>>     myThresholdExceedance.increment(point)
>>> print(myThresholdExceedance.getThresholdExceedance())
[14,14,14,17,12]
>>> print(myThresholdExceedance.getIterationNumber())
50
```

### CODE REVIEW

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+  const Point skewnessEstimator(getSkewness());
+  const UnsignedInteger n = iteration_;
+
+  const Scalar factor1 = n * (n + 1.0) / ((n - 1.0) * (n - 2.0) * (n - 3.0));
+  /*
+  Unbiased estimator:
+  const Scalar factor2 = -3.0 * (n - 1.0) * (n - 1.0) / ((n - 2.0) * (n - 3.0));
+  */
+  /* Corrected with Fischer -3 factor */
+  const Scalar factor2 = -3.0 * (3.0 * n - 5) / ((n - 2.0) * (n - 3.0));
```
Comment (mbaudin47) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:148]: This cannot be correct, because factor2 must converge to -3 when n goes to infinity.

Comment (KieranDelamotte) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:142]: See https://github.com/openturns/openturns/issues/1929 if it helps.

File: python/src/SobolIndicesAlgorithmImplementation_doc.i.in [hunk 2]
```diff
@@ -395,7 +395,7 @@ OT_SobolIndicesAlgorithm_draw_doc
 // ---------------------------------------------------------------------
 
 %define OT_SobolIndicesAlgorithm_DrawCorrelationCoefficients_doc
-"Draw the correlation coefficients. 
+"Draw the correlation coefficients.
```
Comment (mbaudin47) [inline python/src/SobolIndicesAlgorithmImplementation_doc.i.in:398]: This is the doc before the commit:
image
The text should not be in bold. 
This change is unrelated to the current development.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 1 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:142]: Thank you very much.

File: python/src/IterativeMoments_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Notes
+-----
+This class iteratively computes the centered moments of an iteratively increasing
+dataset without storing any data in memory.
+
+See also
+--------
+IterativeAlgorithm"
+
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply python/src/IterativeMoments_doc.i.in:62]: Fixed.

**Comment by mbaudin47** [comment]
I still have to manage special cases with sizes n = 1, n = 2, n = 3, with appropriate code and tests.

**Comment by mbaudin47** [comment]
The ChangeLog has to be updated, which can be done at release time, following @jschueller advice.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 1 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:138]: For your information, I added a detailed explanation of the equation in #1929.

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   *
+   * The object has the default name but it does not
+   * use storage for it.
+   */
+
+  /** Default constructor */
+  IterativeAlgorithmImplementation();
+
+  /** Default constructor */
+  IterativeAlgorithmImplementation(const UnsignedInteger dimension);
```
Comment (jschueller) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:54]: use explicit  keyword

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithm.hxx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+   * The constructor sets a new Id to the object,
+   * so it can be later referenced by a Study object.
+   * It is also declared visible if member of a study.
+   *
+   * The object has the default name but it does not
+   * use storage for it.
+   */
+  IterativeAlgorithm();
+
+  IterativeAlgorithm(const UnsignedInteger dimension);
```
Comment (jschueller) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithm.hxx:52]: use "explicit" keyword

**Review by jschueller — APPROVED** [review]
+1 pending minor edits:
- you need to update the changelog with new classes and features, my remark about not updating the changelog was only about bugfixes, sorry if that wasnt clear (we add new classes less frequently so conflicts are less likely than bugfixes)
- a rebase is needed to hide testcode commits

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx (hunk 4 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:54]: Ok. Fixed.

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithm.hxx (hunk 5 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/openturns/IterativeAlgorithm.hxx:52]: Ok. Fixed.

File: ChangeLog [hunk 6]
```diff
@@ -16,6 +16,9 @@
  * HSICEstimatorTargetSensitivity
  * HSICUStat
  * HSICVStat
+ * IterativeExtrema
+ * IterativeMoments
+ * IterativeThresholdExceedance
```
Comment (jschueller) [inline ChangeLog:21]: add a line in the "major changes" section too

Comment (mbaudin47) [reply ChangeLog:22]: Oups : of course. Done.

**Comment by mbaudin47** [comment]
@KieranDelamotte : Your review would be highly appreciated.

File: lib/src/Base/IterativeStat/IterativeAlgorithm.cxx [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+ * The object has the default name but it does not
+ * use storage for it.
+ */
+IterativeAlgorithm::IterativeAlgorithm()
+  : TypedInterfaceObject<IterativeAlgorithmImplementation>(new IterativeAlgorithmImplementation())
+{
+  // Nothing to do
+}
+
+/** Constructor */
```
Comment (KieranDelamotte) [inline lib/src/Base/IterativeStat/IterativeAlgorithm.cxx:43]: Remove the Doxygen syntax in cxx files, only in hxx ones.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+Point IterativeExtrema::getMax() const
+{
+  return maxData_;
+}
+
+/* Increment method from a Point */
+void IterativeExtrema::increment(const Point & newData)
+{
+  if (newData.getSize() != dimension_) throw InvalidArgumentException(HERE) << "Error: the given Point is not compatible with the dimension of the iterative extrema.";
+  iteration_ += 1;
```
Comment (KieranDelamotte) [inline lib/src/Base/IterativeStat/IterativeExtrema.cxx:91]: Better is ++ iteration_

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+  /* Special case for a size 3 */
+  if (iteration_ == 3) return Point(dimension_, 0.0);
+
+  Point result(dimension_);
+  const Point varianceEstimator(getVariance());
+  const Point skewnessEstimator(getSkewness());
+  const UnsignedInteger n = iteration_;
+
+  const Scalar factor1 = n * (n + 1.0) / ((n - 1.0) * (n - 2.0) * (n - 3.0));
+  const Scalar factor2 = -3.0 * (3.0 * n - 5) / ((n - 2.0) * (n - 3.0));
```
Comment (KieranDelamotte) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:138]: You may add a comment in order to explain the -3.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+  Point coeficientOfVariationData(dimension_, 0.0);
+  const Point meanData(getMean());
+  const Point varData(getVariance());
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    coeficientOfVariationData[i] = std::pow(varData[i], 0.5) / meanData[i];
+  }
+
+  return coeficientOfVariationData;
```
Comment (KieranDelamotte) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:225]: coefficient instead of coeficient

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_extrema.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import openturns as ot
+import openturns.viewer as otv
+
+# %%
+# We first create a one-dimensional Uniform random variable to generate data.
+dim = 1
+distNormal = ot.Uniform()
+
+# %%
+# The :class:`~openturns.IterativeExtrema` class needs to know the dimension of the sample (here 1):
```
Comment (KieranDelamotte) [inline python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_extrema.py:19]: needs the dimension

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_moments.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
+"""
+Using iterative statistics
```
Comment (KieranDelamotte) [inline python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_moments.py:2]: 'Estimate statistics iteratively' to be coherent with the rest.

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_threshold.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
+"""
+Iterative threshold exceedance
```
Comment (KieranDelamotte) [inline python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_threshold.py:2]: Estimate a threshold exceedance iteratively

File: python/src/IterativeThresholdExceedance_doc.i.in [hunk 14]
```diff
@@ -0,0 +1,70 @@
+%feature("docstring") OT::IterativeThresholdExceedance
+"Iterative minimum and maximum.
```
Comment (KieranDelamotte) [inline python/src/IterativeThresholdExceedance_doc.i.in:2]: Iterative threshold exceedance

**Review by KieranDelamotte — APPROVED** [review]
Here are my few comments. Otherwise it is OK for me.

**Comment by jschueller** [comment]
@regislebrun or @sofianehaddad thoughts on this one ?

File: lib/src/Base/IterativeStat/IterativeAlgorithm.cxx (hunk 7 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeAlgorithm.cxx:43]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 8 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:91]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 10 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:225]: Ok. Fixed.

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_extrema.py (hunk 11 above)
Comment (mbaudin47) [reply python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_extrema.py:19]: Ok. Fixed.

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_moments.py (hunk 12 above)
Comment (mbaudin47) [reply python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_moments.py:2]: Ok. Fixed.

File: python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_threshold.py (hunk 13 above)
Comment (mbaudin47) [reply python/doc/examples/numerical_methods/iterative_statistics/plot_iterative_threshold.py:2]: Done.

File: python/src/IterativeThresholdExceedance_doc.i.in (hunk 14 above)
Comment (mbaudin47) [reply python/src/IterativeThresholdExceedance_doc.i.in:2]: Fixed.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 9 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:138]: I was not sure of what comment to write, so I suppressed them completely. I looked on the API manual of Sample.computeKurtosis and found the following explanation:

image

So I thought that the calculation was unclear and performed the calculation presented at:

https://github.com/openturns/openturns/issues/1929

All in all, we may:

- add a small comment in the C++ like "Unbiased estimator" but that this may suggest that other estimators are biased (but they are unbiased),
- add to the C++ comment a reference to [^Joanes1998]
- create a new theory entry which would explain the maths on iterative statistics with, e.g., a reference to Pebay's paper and other references as well.

In the latter case, I can write such a page, but this can (and should) be done afterwards. 

[^Joanes1998]: Joanes, Derrick N.; Gill, Christine A. (1998), "Comparing measures of sample skewness and kurtosis", _Journal of the Royal Statistical Society, Series D_, **47** (1): 183–189, 

What option do you suggest?

Comment (KieranDelamotte) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:138]: The first and third options are good to me. A small comment in the cxx allows for quick comparison with the `Sample` class for developpers.  I left comments with 'unbiased' in the initial classes.
The theory page is a good idea though not related to this specific PR.

Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:148]: I updated the comments and added a reference to Joanes's reference. I hope it fits your requirements.

Comment (KieranDelamotte) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx:148]: That's fine.

**Comment by mbaudin47** [comment]
I suggest a final review from @regislebrun or @sofianehaddad so that we may see if there is any fix left before merging.

**Comment by mbaudin47** [comment]
I kindly suggest a review from @regislebrun and @sofianehaddad or an approval.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(IterativeExtrema)
+
+static const Factory<IterativeExtrema> Factory_IterativeExtrema;
+
+/* Constructor */
+IterativeExtrema::IterativeExtrema(const UnsignedInteger dimension)
+  : IterativeAlgorithmImplementation(dimension)
+  , minData_(dimension, 0.0)
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeExtrema.cxx:37]: useless. Points are initialized at zero by default.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 8 above)
Comment (regislebrun) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:91]: Really?

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  }
+  else
+  {
+    for (UnsignedInteger i = 0; i < dimension_; ++i)
+    {
+      maxData_[i] = newData[i];
+      minData_[i] = newData[i];
+    }
+  }
+}
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeExtrema.cxx:114]: So many braces... no comment... When there is only one line in the block braces are optional

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        if (newData(j, i) > maxData_[i])
+        {
+          maxData_[i] = newData(j, i);
+        }
+        if (newData(j, i) < minData_[i])
+        {
+          minData_[i] = newData(j, i);
+        }
+      }
+    }
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeExtrema.cxx:137]: Braces braces braces... and no comment

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+Point IterativeMoments::getCoefficientOfVariation() const
+{
+  if ((orderMax_ < 2)) throw InternalException(HERE) << "Error: unavailable method, the declared maximum order is lower than 2.";
+
+  Point coefficientOfVariationData(dimension_, 0.0);
+  const Point meanData(getMean());
+  const Point varData(getVariance());
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    coefficientOfVariationData[i] = std::pow(varData[i], 0.5) / meanData[i];
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:232]: std::sqrt, better than std::pow

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+Point IterativeMoments::getStandardErrorOfTheMean() const
+{
+  if ((orderMax_ < 2)) throw InternalException(HERE) << "Error: unavailable method, the declared maximum order is lower than 2.";
+
+  Point standardErrorOfTheMean(dimension_, 0.0);
+  const Point varData(getVariance());
+
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    standardErrorOfTheMean[i] = std::pow(varData[i], 0.5) / std::pow(iteration_, 0.5);
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:248]: std::sqrt, better than std::pow, and take a unique square root of a ratio instead of the ratio of two square roots

File: lib/src/Base/IterativeStat/IterativeMoments.cxx [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+Point IterativeMoments::getStandardDeviation() const
+{
+  if ((orderMax_ < 2)) throw InternalException(HERE) << "Error: unavailable method, the declared maximum order is lower than 2.";
+
+  Point standardDeviationData(dimension_, 0.0);
+  const Point varData(getVariance());
+
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    standardDeviationData[i] = std::pow(varData[i], 0.5);
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeMoments.cxx:264]: std::sqrt

File: lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+/* Increment the internal data with a Point */
+void IterativeThresholdExceedance::increment(const Point & newData)
+{
+  if (newData.getSize() != dimension_) throw InvalidArgumentException(HERE) << "Error: the given Point is not compatible with the dimension of the iterative threshold exceedance.";
+  ++ iteration_;
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    if (newData[i] > thresholdValue_)
+    {
+      data_[i] += 1.0;
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx:88]: See my braces

File: lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+    ++ iteration_;
+    for (UnsignedInteger i = 0; i < dimension_; ++i)
+    {
+      if (newData(j, i) > thresholdValue_)
+      {
+        data_[i] += 1.0;
+      }
+    }
+  }
+}
```
Comment (regislebrun) [inline lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx:109]: count my braces

**Review by regislebrun — APPROVED** [review]
Minor remarks, otherwise good to go!

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 15 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:37]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 8 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:91]: Oups. There was two of them. Fixed.

Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:91]: By the way, I reimplemented the body of increment(sample), so that it loops over its points and call increment(point). This factors the code.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 16 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx:108]: I agree: I commented all for loops, and removed unnecessary braces where appropriate.

File: lib/src/Base/IterativeStat/IterativeExtrema.cxx (hunk 17 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeExtrema.cxx]: Ok, fixed.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 19 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 20 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx (hunk 21 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx (hunk 22 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeThresholdExceedance.cxx:94]: Ok. Fixed.

File: lib/src/Base/IterativeStat/IterativeMoments.cxx (hunk 18 above)
Comment (mbaudin47) [reply lib/src/Base/IterativeStat/IterativeMoments.cxx]: Ok. Fixed.

**Review by regislebrun — APPROVED** [review]
